File: saksflyt/src/test/java/no/nav/melosys/saksflyt/steg/ufm/OpprettSedDokumentTest.java
package no.nav.melosys.saksflyt.steg.ufm;

import java.time.LocalDate;
import java.util.Collections;

import no.nav.melosys.domain.Behandling;
import no.nav.melosys.domain.ProsessDataKey;
import no.nav.melosys.domain.ProsessSteg;
import no.nav.melosys.domain.Prosessinstans;
import no.nav.melosys.domain.dokument.sed.BucType;
import no.nav.melosys.domain.dokument.sed.SedDokument;
import no.nav.melosys.domain.dokument.sed.SedType;
import no.nav.melosys.domain.eessi.melding.MelosysEessiMelding;
import no.nav.melosys.domain.eessi.melding.Periode;
import no.nav.melosys.domain.eessi.melding.Statsborgerskap;
import no.nav.melosys.domain.kodeverk.lovvalgsbestemmelser.Lovvalgbestemmelser_883_2004;
import no.nav.melosys.repository.SaksopplysningRepository;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.verify;

@RunWith(MockitoJUnitRunner.class)
public class OpprettSedDokumentTest {

    @Mock
    private OpprettSedDokumentFelles opprettSedDokumentFelles;

    private OpprettSedDokument opprettSedDokument;

    @Before
    public void setup() {
        opprettSedDokument = new OpprettSedDokument(opprettSedDokumentFelles);
    }

    @Test
    public void utfoerSteg() throws Exception {

        SedDokument sedDokument = new SedDokument();
        sedDokument.setLovvalgBestemmelse(Lovvalgbestemmelser_883_2004.FO_883_2004_ART12_1);
        sedDokument.setBucType(BucType.LA_BUC_04);
        sedDokument.setSedType(SedType.A009);

        Prosessinstans prosessinstans = new Prosessinstans();
        prosessinstans.setData(ProsessDataKey.EESSI_MELDING, hentMelosysEessiMelding());
        prosessinstans.setBehandling(new Behandling());

        opprettSedDokument.utfør(prosessinstans);

        verify(opprettSedDokumentFelles).opprettSedSaksopplysning(any(MelosysEessiMelding.class), any(Behandling.class));
        assertThat(prosessinstans.getSteg()).isEqualTo(ProsessSteg.REG_UNNTAK_HENT_PERSON);
    }

    private MelosysEessiMelding hentMelosysEessiMelding() {
        MelosysEessiMelding melding = new MelosysEessiMelding();
        melding.setAktoerId("123");
        melding.setArtikkel("12_1");
        melding.setDokumentId("123321");
        melding.setGsakSaksnummer(432432L);
        melding.setJournalpostId("j123");
        melding.setLovvalgsland("SE");

        Periode periode = new Periode();
        periode.setFom(LocalDate.of(2012, 12, 12));
        periode.setTom(LocalDate.of(2012, 12, 13));
        melding.setPeriode(periode);

        Statsborgerskap statsborgerskap = new Statsborgerskap();
        statsborgerskap.setLandkode("SE");

        melding.setRinaSaksnummer("r123");
        melding.setSedId("s123");
        melding.setStatsborgerskap(
            Collections.singletonList(statsborgerskap));
        melding.setSedType("A009");
        melding.setBucType("LA_BUC_04");
        return melding;
    }
}

File: saksflyt/src/test/kotlin/no/nav/melosys/saksflyt/steg/sed/OpprettSedDokumentTest.kt
package no.nav.melosys.saksflyt.steg.sed

import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.mockk
import io.mockk.verify
import no.nav.melosys.domain.Behandling
import no.nav.melosys.domain.eessi.Periode
import no.nav.melosys.domain.eessi.melding.MelosysEessiMelding
import no.nav.melosys.domain.eessi.melding.Statsborgerskap
import no.nav.melosys.domain.forTest
import no.nav.melosys.saksflytapi.domain.ProsessDataKey
import no.nav.melosys.saksflytapi.domain.Prosessinstans
import no.nav.melosys.saksflytapi.domain.forTest
import no.nav.melosys.service.saksopplysninger.OpprettSedDokumentService
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import java.time.LocalDate

@ExtendWith(MockKExtension::class)
class OpprettSedDokumentTest {

    @MockK
    private lateinit var opprettSedDokumentService: OpprettSedDokumentService

    private lateinit var opprettSedDokument: OpprettSedDokument

    @BeforeEach
    fun setup() {
        opprettSedDokument = OpprettSedDokument(opprettSedDokumentService)
    }

    @Test
    fun utfoerSteg() {
        every { opprettSedDokumentService.opprettSedSaksopplysning(any<MelosysEessiMelding>(), any<Behandling>()) } returns mockk()

        val prosessinstans = Prosessinstans.forTest {
            medData(ProsessDataKey.EESSI_MELDING, hentMelosysEessiMelding())
        }
        prosessinstans.behandling = Behandling.forTest { }


        opprettSedDokument.utfør(prosessinstans)


        verify { opprettSedDokumentService.opprettSedSaksopplysning(any<MelosysEessiMelding>(), any<Behandling>()) }
    }

    private fun hentMelosysEessiMelding() = MelosysEessiMelding().apply {
        aktoerId = "123"
        artikkel = "12_1"
        dokumentId = "123321"
        gsakSaksnummer = 432432L
        journalpostId = "j123"
        lovvalgsland = "SE"
        periode = Periode().apply {
            fom = LocalDate.of(2012, 12, 12)
            tom = LocalDate.of(2012, 12, 13)
        }
        rinaSaksnummer = "r123"
        sedId = "s123"
        statsborgerskap = listOf(Statsborgerskap("SE"))
        sedType = "A009"
        bucType = "LA_BUC_04"
    }
}
